Elly Vendingmachine step4

dely2p opened the pull request:
Vendingmachine step4 입니다.

dely2p committed:
Detach the OutputView as Admin and User #23

OutputView를 사용자와 관리자로 분리해서 출력하도록 수정했다.

dely2p committed:
add printMachineModeMenu method in output #24

관리자와 사용자를 선택할 수 있는 메뉴출력을 추가했다.

dely2p committed:
Combination of isolated OutputView class #24

사용자와 관리자로 분리한 OutputView 클래스를 main에서 조합해서 사용했다.

dely2p committed:
fix code to add function of Select Administrator and User Functions #25

관리자기능과 사용자기능 선택을 위해 메소드 추가 및 수정

godrm commented on VendingMachine/VendingMachine/AdminOutputView.swift:
```diff
@@ -0,0 +1,56 @@
+            print("get Beverage error")
+        }
+        return Beverage(brand: "", weight: 0, price: 0, name: "", manufactureDate: Date())
+    }
```
여기 beverageFromMenuNumber() 함수는 AdminOutputView 역할일까요?
저라면 Beverage객체를 만들어주는 빌더, 팩토리 객체를 따로 만들었을 것 같습니다.

godrm commented on VendingMachine/VendingMachine/AdminOutputView.swift:
```diff
@@ -0,0 +1,56 @@
+            let numberOfBeverage = menu.value.count
+            print("\(index+1)] \(menu.value[0].kindOf)(\(numberOfBeverage)개) ")
+        }
+        print()
```
[ObjectIdentifier : [Beverage]] 형태를 대신할 객체가 필요하지는 않을까요?

godrm commented on VendingMachine/VendingMachine/UserOutputView.swift:
```diff
@@ -0,0 +1,56 @@
+            }
+        }
+        throw InventoryBox.VendingMachinError.invalidBeverage
+    }
```
이 부분은 AdminOutputView랑 동일한거 아닌가요?

godrm commented on VendingMachine/VendingMachine/main.swift:
```diff
@@ -11,55 +11,90 @@ import Foundation
+                    case 1:
+                        vendingMachine.addInInventory(beverageName: choiceBeverageName, number: inputValue[2])
+                        adminOutputView.printAddInventory(choiceBeverage: choiceBeverageName, number: inputValue[2])
+                    case 2:
```
여기 switch-case 구문에서 사용하는 Int 타입 대신에 메뉴에 해당하는 의미있는 코드로 사용하세요. 항상 하드코딩하는 숫자값, 문자열은 코드로 대체하도록 하세요.

godrm commented on VendingMachine/VendingMachine/main.swift:
```diff
@@ -11,55 +11,90 @@ import Foundation
+            } while isAdminModeContinue
+        // 사용자 모드
+        }else if mode == 2 {
+            repeat {
```
mode == 1 과 2 경우를 메소드를 분리하는게 어떨까요? 
OutputView를 나눈것처럼 Controller를 구분하는 것도 방법인 것 같습니다.

godrm reviewed (commented):
고민한 내용들이 잘 표현되어 있네요.
다음 피드백을 확인해서 개선해보세요.

dely2p committed:
create method to create a Beverage object #27

Beverage 객체를 만들어주는 메소드를 InventoryBox에 생성했다.

dely2p committed:
separated in UserContoller and AdminController #30

관리자모드와 사용자모드를 각 Contoller Class로 분리했다.

dely2p committed:
fix outputValue error in InventoryBox Class #30

InventoryBox Class에서 Beverage값이 제대로 넘어오지 않는 에러를 수정했다. enum값을 사용하도록 바꾸면서 에러가 생겼던 것 같다.

dely2p committed:
fix error in UserController #30

사용자 컨트롤러의 vendingMachine이 초기화 되는 에러를 잡았다.

dely2p committed:
fix InventoryMenu enum from description to BeveragekeyFromMenu() #27

enum 안에서 description을 사용하여 ObjectIdentifier를 리턴했던 것을 메소드를 사용하여 리턴하도록 수정했다.

dely2p committed:
add AdminDTO, UserDTO to make data transfer object #29

하드코딩으로 배열의 값에 접근하지 않도록 AdminDTO와 UserDTO를 만들었다.

dely2p committed:
fix choice function to user or admin in main #29

사용자모드와 관리자모드를 구분하는 기능을 Mode enum을 사용하여 수정했다.

dely2p committed:
fix DTO struct, naming #29

기존 UserDTO를 사용하는 것에 따라 BeveragePurchaseDTO와 AddCoinDTO로나누었다.

dely2p committed:
add DTO initialize to use in user, admin InputView #29

user와 admin InputView를 나누고 각각 사용할 DTO를 초기화 했다.

dely2p committed:
fix group folder to separate Model, View, Controller

MVC 각각의 그룹으로 나누었다.

dely2p committed:
Replace with meaningful code #29

인덱스나 하드코딩이 아닌 DTO를 사용하여 의미있는 코드로 바꿨다.

godrm commented on VendingMachine/VendingMachine/View/AdminOutputView.swift:
```diff
@@ -0,0 +1,34 @@
+            let numberOfBeverage = menu.value.count
+            print("\(index+1)] \(menu.value[0].kindOf)(\(numberOfBeverage)개) ")
+        }
+        print()
```
override 한 메소드에서 super. printBeverageMenu()를 사용하지 않는 이유가 있나요?

dely2p replied:
override 한 메소드에서는 출력 형태가 조금 달라서 super.printBeverageMenu()를 사용하지 않습니다:)

godrm replied:
혹시 공통적인 부분을 전혀 활용할 방법은 없었나요? 그렇다면 할 수 없겠네요.

godrm commented on VendingMachine/VendingMachine/View/UserInputView.swift:
```diff
@@ -0,0 +1,28 @@
+    func separateInputValues(input: String) -> VendingMachineUser {
+        let seperatedValue = input.split(separator: " ").map({Int($0) ?? 0})
+        if seperatedValue[0] == 1 {
+            return AddCoin(amount: seperatedValue[1])
```
와 이건 완전 속았네요. AddCoin() 함수인줄... DTO 이름을 이렇게 만들면 대략난감하네요 ㅎㅎ

godrm commented on VendingMachine/VendingMachine/View/UserOutputView.swift:
```diff
@@ -65,4 +44,13 @@ class OutputView {
+    func getBeverageKey(menuNumber: Int) throws -> ObjectIdentifier {
+        for (index, beverage) in self.availableBeverage.enumerated() {
+            if menuNumber == (index + 1) {
+                return beverage.key
```
이게 순서가 보장이 되나요? self.availableBeverage은 사전이라서 순서가 일정하지 않을텐데요!

dely2p replied:
앗 메뉴판 순서보장 안됩니다^^;; 수정하겠습니다

dely2p replied:
아! self.availableBeverage는 사전이라 값을 가져오는 것의 순서보장은 안되는데 index를 이용해서 순서에 맞는 값을 가져오도록 만들었습니다.

godrm replied:
아마 index도 순서가 보장이 안될텐데??

dely2p replied:
index 순서가 보장되도록 수정했습니다:)
- 이름순으로 정렬하는 메소드(menuOfSerialized())를 만들었습니다.
- 각 outputView에서 print 할 때 menuOfSerialized()를 사용하여 정렬된 Beverage Key값을 사용하도록 만들었습니다.
- InventoryMenu enum을 사용하여 메뉴 번호 입력시 해당 Beverage Data를 리턴하도록 했습니다.

dely2p committed:
fix DTO name from AddCoin to CoinBox #32

AddCoin이라는 DTO 이름이 함수로 오해할 가능성이 크므로 CoinBox로 이름을 수정했다.

dely2p committed:
fix print menu to ensure menu order is always maintained #33

메뉴판이 항상 순서보장되도록 바꾸었다.

dely2p committed:
fix switch structure in repeate-while part to go first menu

관리자모드와 사용자모드에서 0번을 입력하면 초기 모드 선택으로 돌아갈 수 있도록 수정했다.

godrm commented on VendingMachine/VendingMachine/View/UserOutputView.swift:
```diff
@@ -11,8 +11,9 @@ import Foundation
-            print("\(index+1))\(menu.value[0].kindOf) \(menu.value[0].price)(\(menu.value.count)개)")
+        let serializedBeverage = menuOfSerialized()
+        for (index, beverageKey) in serializedBeverage.enumerated() {
+            print("\(index+1))\(availableBeverage[beverageKey]?.first?.kindOf ?? "") \(availableBeverage[beverageKey]?.first?.price ?? 0)(\(availableBeverage[beverageKey]?.count ?? 0)개)")
```
availableBeverage 가 멤버 변수로 있는데 매개변수가 또 있네요? 뭐가 다른걸까요?

godrm commented on VendingMachine/VendingMachine/View/OutputView.swift:
```diff
@@ -26,4 +27,18 @@ class OutputView {
+                    break
+                }
+            }
+        }
```
여기서 이중 반복문으로 비교하지 않고도
`availableBeverage.map({$0.value.first!.kindOf}).sorted(by: <)`  형태로 우선 $0.value.first로 map하고 sorted 할 때 kindOf 기준으로 비교하고 다시 ObjectIdentifier를 구해서 map 하면 어떨까요? 이건 꼭 해봅시다.

dely2p replied:
오! 코드가 짧아졌어요!! >_<
sort()와 map()을 사용해서 한 줄로 수정했습니다.

dely2p committed:
Double for Door Code reduced #35

이중 for문을 sort()와 map()을 이용해서 한 줄로 만들었다.

dely2p committed:
clear the availableBeverage member parameter and accept the value by parameter #34

availableBeverage 멤버변수를 지우고 파라미터로 들어온 전체 Beverage를 사용하도록 수정했다.

godrm commented:
깔끔하게 정리했군요. 마무리합시다. 
수고많으셨습니다.
